### PR TITLE
🚀 Rework unesc for a 63+% performance boost to all of postcss.

### PR DESCRIPTION
**Pull request by samccone**
In profiling postcss I found that a significant amount of time was being
spent in [`unesc`](https://github.com/postcss/postcss-selector-parser/commits/master/src/util/unesc.js), this was due to the expensive regex checks that were
being performed on the fly for every selector in the codebase which looked to be performing quite poorly inside of modern node and v8.

## Old

image

----

As an experiment and based on some prior experience with this class of slowdown I migrated the implementation to one that performs a scan through the string instead of running a regex replace. By testing this on my local application I instantly saw the work from this function go from > 900 ms to ~100ms.

## New

image

This implementation passes all of the existing test cases and aims to mirror the prior implementation's implementation details :)

-----

Based on my application I am seeing the major wins come from purgecss dropping my total application build by multiple seconds! 🔥 

---------

## Perf testing

I set up a simple perf test here:
https://github.com/samccone/postcss-selector-parser/commit/be99c239a394327c7788abfbccc021821f26a0bf which takes all of tailwind's css selectors and then unescapes them 100 times.

(left old, right new) 
image

__Previously this simple test took **20.17ms** now it takes **7.1ms** -- a 63% difference__ 

Finally based on input I did a comparison between the new, old, and chrome's implementation for this work

```
original 
        avg: 4.765ms
        std: 2.5159ms
        max: 35ms
        min: 3ms
        
new 
        avg: 1.76ms
        std: 0.7915ms
        max: 12ms
        min: 1ms
        
chrome 
        avg: 4.383ms
        std: 1.7602ms
        max: 34ms
        min: 3ms
```

You can run the test case here https://github.com/samccone/postcss-selector-parser/blob/sjs/perf/perf/index.js, the testing is similar to what I did before but now I run the test cases 1000 times each.

Finally I tested this on tailwind and saw the wins propagating to the ecosystem with this patch!
https://twitter.com/samccone/status/1381423236253032454

### CODE REVIEW

File: src/util/unesc.js [hunk 1]
```diff
@@ -1,19 +1,96 @@
...
+
+const CSS_SPECIAL_CHARS = new Set([
+    "!", "\"", "#", "$", "%", "&", "'", "(", ")", "*", "+", ",", "-", ".", "/", ";", "<", "=", ">", "?", 
+    "@", "[", "]", "^", "`", "{", "|", "}", "~", "_", ":",
+]);
+
+const CSS_SPECIAL_CHARS_HEX_LONG = {}; 
+const CSS_SPECIAL_CHARS_HEX_SHORT = {};
+
+for (const char of Array.from(CSS_SPECIAL_CHARS)) {
```
Comment (mathiasbynens) [inline src/util/unesc.js:12]: Sets are iterable, so we can even avoid casting to array here:

```suggestion
for (const char of CSS_SPECIAL_CHARS) {
```

File: src/util/unesc.js [hunk 2]
```diff
@@ -1,19 +1,96 @@
...
+
+            // Retain a pair of \\ if double escaped `\\\\`
+            // https://github.com/postcss/postcss-selector-parser/commit/268c9a7656fb53f543dc620aa5b73a30ec3ff20e
+            if (str[i +1] === "\\") {
+                ret += "\\";
+                i++;
+                continue;
+            }
+
+            // Need to check if we have an emoji
```
Comment (mathiasbynens) [inline src/util/unesc.js:76]: I might be misunderstanding, but I think the intention here is to detect astral symbols (i.e. code points above U+FFFF, i.e. those in the range from U+010000 to U+10FFFF). This is orthogonal to the use of emoji. (Some emoji have a code point `<` U+FFFF.) Did I get that right?

In that case, there could be up to 6 hex digits (and that's [the maximum](https://drafts.csswg.org/css-syntax/#consume-escaped-code-point) that CSS allows as well).

File: src/util/unesc.js [hunk 3]
```diff
@@ -1,19 +1,96 @@
...
+// Many thanks for this post which made this migration much easier.
+// https://mathiasbynens.be/notes/css-escapes
+
+const CSS_SPECIAL_CHARS = new Set([
+    "!", "\"", "#", "$", "%", "&", "'", "(", ")", "*", "+", ",", "-", ".", "/", ";", "<", "=", ">", "?", 
+    "@", "[", "]", "^", "`", "{", "|", "}", "~", "_", ":",
+]);
+
+const CSS_SPECIAL_CHARS_HEX_LONG = {}; 
+const CSS_SPECIAL_CHARS_HEX_SHORT = {};
```
Comment (mathiasbynens) [inline src/util/unesc.js:10]: These seem to be used as maps, so we could use actual `Map`s. It might even speed up this patch even further, since during the lookups (e.g. line 26) we’d then be avoiding prototype chain walks for any misses.

**Review by mathiasbynens — COMMENTED** [review]
Impressive investigation + speed-up! I hope you don’t mind I took a look and left some comments with further ideas — WDYT?

**Comment by alexander-akait** [comment]
Great, let's improve it (from feedback above) and do release

**Comment by samccone** [comment]
Thanks @mathiasbynens and @alexander-akait for the review! All updated and sped up another 10% or so!

**Comment by mathiasbynens** [comment]
Btw, I wonder how our DevTools implementation compares: https://source.chromium.org/chromium/chromium/src/+/master:third_party/devtools-frontend/src/front_end/panels/elements/StylesSidebarPane.js;l=3282-3301;drc=fca98ed75dcae230e4a6fb225ae2c46c43d4939b

**Comment by alexander-akait** [comment]
@samccone @mathiasbynens Can we do small benchmarks, interesting how it faster/slow

**Comment by samccone** [comment]
Hi @alexander-akait yes, at the bottom of the original commit message I added a perf section outlining the old and new implementation. 

As far as reusing chrome's implementation with a regex this is possible but we would need to update the license of this project to include chrome's licence as well.  

Finally we can totally do a performance test between chrome's and this implementation. Upon first pass chrome's implementation handles a nice edge case that we do not (invalid unicode), but does not handle a few other decoding cases that the existing test cases require. (Changing this while more correct would be a backwards incompatible change to this core part of postcss which would be a larger undertaking)

I will do some further investigation today! 🚀

**Comment by samccone** [comment]
Hi @alexander-akait and @mathiasbynens a few updates

1. I updated the implementation to now *new* handle the spec edgecase for 0 hex value escaping, long surrogates, and out of bound value https://drafts.csswg.org/css-syntax/#maximum-allowed-code-point - this is the new commit.
2. I ran new perf tests against chrome's implementation and found that our implementation in this CL is significantly faster. 

The numbers are here:

```
original 
        avg: 4.765ms
        std: 2.5159ms
        max: 35ms
        min: 3ms
        
new 
        avg: 1.76ms
        std: 0.7915ms
        max: 12ms
        min: 1ms
        
chrome 
        avg: 4.383ms
        std: 1.7602ms
        max: 34ms
        min: 3ms
```
        
And the testing methodology can be viewed here https://github.com/samccone/postcss-selector-parser/blob/sjs/perf/perf/index.js#L160-L185

**Review by alexander-akait — APPROVED** [review]
/cc @mathiasbynens Can you look too?  It would be nice if a couple more eyes looked at the code, thanks

File: src/__tests__/util/unesc.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    t.deepEqual(tree.nodes[0].nodes[0].value, 'édition');
+});
+
+test('class selector with escaping', '.\\1D306', (t, tree) => {
+    t.deepEqual(tree.nodes[0].nodes[0].value, '𝌆');
+});
+
+test('class selector with escaping with more chars', '.\\1D306k', (t, tree) => {
+    t.deepEqual(tree.nodes[0].nodes[0].value, '𝌆k');
+});
```
Comment (mathiasbynens) [inline src/__tests__/util/unesc.js:46]: Test that trailing whitespace gets consumed even in the case of an astral character:

```suggestion
test('class selector with escaping with more chars', '.\\1D306k', (t, tree) => {
    t.deepEqual(tree.nodes[0].nodes[0].value, '𝌆k');
});

test('class selector with escaping with more chars', '.foo\\1D306 bar', (t, tree) => {
    t.deepEqual(tree.nodes[0].nodes[0].value, 'foo𝌆bar');
});
```

File: src/util/unesc.js [hunk 5]
```diff
@@ -1,19 +1,82 @@
...
+            const gobbled = gobbleHex(str.slice(i+1, i+7));
+            if (gobbled !== undefined) {
+                ret += gobbled[0];
+                i += gobbled[1];
+                continue;
+            }
+
+            // Retain a pair of \\ if double escaped `\\\\`
+            // https://github.com/postcss/postcss-selector-parser/commit/268c9a7656fb53f543dc620aa5b73a30ec3ff20e
+            if (str[i +1] === "\\") {
```
Comment (mathiasbynens) [inline src/util/unesc.js:64]:
```suggestion
            if (str[i + 1] === "\\") {
```

**Review by mathiasbynens — APPROVED** [review]

File: src/__tests__/util/unesc.js (hunk 4 above)
Comment (samccone) [reply src/__tests__/util/unesc.js:46]: added this case!

File: src/util/unesc.js (hunk 5 above)
Comment (samccone) [reply src/util/unesc.js]: added.

**Comment by schuay** [comment]
To add some history, slow String.p.replace with global regexp and a callable replacer has been known for a long time, and unfortunately I haven't gotten around to fixing it yet - apologies for that. The previous situation was in fact in a very similar context, see https://github.com/nodejs/node/pull/16986#discussion_r150728733 and crbug.com/v8/7081.

I'm hoping to spend some quality time with regexp later this year. For priorization, it would help if you file a V8 bug report for this, just to point out that this is a real issue, still impacting real cases.

In the meantime, it makes sense to use perf workarounds. I know Leszek has been looking at one based on `RegExp.p.exec` (exec, unlike replace, is actually very fast in the common case), which I suppose he will post here soon.

**Comment by LeszekSwirski** [comment]
Right, the following:
```
function regexpexec(str) {
    let new_str = "";
    let prev_index = 0;
    let match;
    while ((match = unescapeRegExp.exec(str)) != null) {
        let escaped = match[1];
        let escapedWhitespace = match[2];
        if (match.index > prev_index) {
            new_str += str.slice(prev_index, match.index);
        }

        const high = "0x" + escaped - 0x10000;

        // NaN means non-codepoint
        // Workaround erroneous numeric interpretation of +"0x"
        // eslint-disable-next-line no-self-compare
        new_str += high !== high || escapedWhitespace
            ? escaped
            : high < 0
            ? // BMP codepoint
              String.fromCharCode(high + 0x10000)
            : // Supplemental Plane codepoint (surrogate pair)
              String.fromCharCode(
                  (high >> 10) | 0xd800,
                  (high & 0x3ff) | 0xdc00
              );
        prev_index = unescapeRegExp.lastIndex;
    }
    if (str.length > prev_index) {
        new_str += str.slice(prev_index, str.length);
    }
    return new_str;
}
```

gives me a 50% speedup over this manual string walk:

```
original
        avg: 2.592ms
        std: 0.5528ms
        max: 8ms
        min: 2ms

new
        avg: 1.578ms
        std: 0.5531ms
        max: 7ms
        min: 1ms

chrome
        avg: 2.872ms
        std: 0.4956ms
        max: 8ms
        min: 2ms

Regexp.exec
        avg: 1.002ms
        std: 0.1095ms
        max: 2ms
        min: 0ms
```

**Comment by samccone** [comment]
@LeszekSwirski unfortunately the solution you listed does not pass all of the test cases added in this change. This is not reflective of your solution being incorrect but rather is more reflective of the existing implementation having multiple bugs that were uncovered and fix during this code-review process.

At this point I see several options

## 1. ignore the fixes to spec bugs that this change fixes and take the solution you wrote 

(pros) 
```
+ much faster than what is in place
```
(cons) 

```
- carries forward the incorrect behavior of unesc forward
```

## 2. Take the current implementation 

(pros) 
```
+ much faster than what is in place 
+ more spec compliant and correct
```

(cons)
```
- not as fast as option 1
```

## 3. (maybe 2.a) follow option 2 and follow up with further perf improvements

(pros)

```
+ helpful to the ecosystem
+ much faster than what is currently released
+ more spec compliant
```

(cons)

```
- we know there is still room for improvement through further refactoring
```

---

@alexander-akait my personal recommendation would be to folllow option 3 with continued community investment to continue to iterate on the optimal solution. 

Please all let me know how you would like to move forward.

**Comment by samccone** [comment]
@schuay v8 bug filed! https://bugs.chromium.org/p/v8/issues/detail?id=11664

**Comment by alexander-akait** [comment]
Big thanks

**Comment by LeszekSwirski** [comment]
@samccone Makes sense, although fixing that V8 performance issue of course won't help with the correctness issues you found here.
